src/cocotb/share/lib/vpi/VpiImpl.cpp

```diff
@@ -169,6 +170,11 @@ static gpi_objtype_t const_type_to_gpi_objtype(int32_t const_type) {
         case vpiBinaryConst:
         case vpiOctConst:
         case vpiHexConst:
+            if (vpi_get(vpiSize, hdl) <= 32) {
+                return GPI_INTEGER;
+            } else {
+                return GPI_NET;
+            }
```

Comment:
Shouldn't this always be `GPI_NET`? Or are we making a bad assumption somewhere else?

Reply:
I think there are probably several issues with integer nets/var/parameters.

For one, `IntegerObject` (mapped to `GPI_INTEGER`) assumes 32-bits when checking overflow:
https://github.com/cocotb/cocotb/blob/d6c1a2ec9fa3e41d071c8e5ee2e9965dd9295a7d/src/cocotb/handle.py#L941

Reply:
Yeah I really don't like vpiConst types, I don't think they make any sense. The subtypes for that seem to more accurately describe the format of literals in the CST, which aren't useful to the sim. And not every parameter is a literal. Do you think we can change these back on verilator, and instead use the vpiConstantVariable property to mark them as constant? These technically aren't variables, but I think that's a much better way than this for describing params @marlonjames

Reply:
I had the same thought with `vpiConstantVariable`. I think we should be using that when we can, but I'm going out on a limb and assuming simulator support is non-existant. It might be worth looking to see if any simulators support it. And I wonder if there's a way to look at the parameter type name so we can map everything perfectly for the handful of types listed below by Marlon.

What we want to see is 4 value types (which includes `integer`) map to logic and 2-value types to ints.

Reply:
I think most simulators are going to return type as `vpiParameter`, not a variable, so that property won't work for those.

Reply:
@ktbarrett parameter objects have a `vpiTypespec` association that would give you the info, but I bet support for that is even lower than everything else discussed.

Reply:
Thinking about it, I think the `vpiConstType` maps to the `s_vpi_value.format` for retrieving the value.
`vpiDecConst` => `vpiDecStrVal`,
`vpiBinaryConst` => `vpiBinStrVal`,
`vpiOctConst` => `vpiOctStrVal`,
`vpiHexConst` => `vpiHexStrVal`,
`vpiIntConst` => `vpiIntVal`

@ktbarrett I don't know how to map 4-state vs 2-state unless we can get the Typespec info.
I'm going to write up a small VPI program so we can see what info simulators are giving for different parameters.

Reply:
`vpiConstType` is probably a misleading name. I think Andrew is correct that I think it's more about the format of the *literal* used to initialized the parameter, and less about the actual type of the parameter, which is what we care about.

Reply:
For the edification of anybody reading this in the future, you can use the `int_param_check` VPI program from https://github.com/marlonjames/vpi-experiments to get information about `vpiConstType` for your simulator of choice.